src/common/libsubprocess/fbuf.c

```diff
@@ -32,6 +32,8 @@ struct fbuf {
     int buflen;
     fbuf_notify_f notify_cb;
     void *notify_cb_arg;
+    fbuf_credit_f credit_cb;
+    void *credit_cb_arg;
```

Comment:
An observation is that the original notify callback could be modified so that it is also called whenever `cbuf_used()` changes.  This would not break the existing users (ev_fbuf_read and ev_fbuf_write).

In fact it would allow ev_fbuf_write to handle this "event" as well.  Wouldn't that be more in keeping with the existing design than having the new fbuf callback invoke the subprocess `on_credit()` callback directly?

IOW arrange for the the fbuf_write_watcher to have a FLUX_POLLOUT event when the value of `fbuf_bytes()` has changed because buffer data has been written.  Then instead of having separate `local_in_cb()` and `local_in_credit_cb()` callbacks (one a reactor watcher and one not), just have the `local_in_cb()` reactor watcher.

I think this would be simpler and more consistent with the original design, and thus less confusing?

Reply:
hmmm.  I see where you're coming from.  I think it is plausible for this work.  However, I'm not sure the overall changes would make it worthwhile.  

The main thing is the fbuf notify callback is currently an "edge" callback, only getting called on "full -> not full" and "empty -> not empty" transitions.  If we convert it to handle byte changes, we need to change that callback to deal with "before" and "after" byte changes.  Worst case, callback function prototype changes everywhere.  Or more annoying case, have to keep track of how many bytes were previously in the buffer and do math after calling `fbuf_bytes()` to see current usage.

Lemme think about this more and see if there's a simpler solution.  But it seems like there'd be a ton of fallout as a result.

Reply:
When I was thinking about it earlier I tried it to see what broke.  That attempt is here FWIW:  db1214a1fde4c04c8039030ebd7fe8e32d32c83d

Reply:
Yeah, I don't think `fbuf` would change that much but ...

```
static void fbuf_notify (struct fbuf *fb, int old_used)
{
    if (cbuf_used (fb->cbuf) != old_used) {
        if (fb->cb)
            fb->cb (fb, fb->cb_arg);
    }
}
```

B/c we're not passing the old and/or new bytes to the callback, something has to be done there to inform the callback what happened.  So that's what I'm thinking could really change things.

Reply:
as I slept on this last night, I began moving towards "it's a lot of work, but may be worth it" to move towards this approach.  I think everything is contained within `libsubprocess` and won't bleed outside of it.  I think minimally, lets give it a shot.

Reply:
Let me know if I can help.  Generally, I think the main role of the watcher is to trigger a callback when the reactor would otherwise sleep, and the "action" should be something taken by the callback itself.  So probably the subprocess channel  sttruct should track the available space in the buffer and the watcher callback should just check if that has changed.